metasrv/src/watcher/watcher_manager.rs

```diff
@@ -0,0 +1,404 @@
+
+            let set = self.watcher_range_set.get_by_point(&kv.key);
+            for range_key in set.iter() {
+                let watcher_id = range_key.key.id;
+                let filter = range_key.key.filter;
+
+                // filter out event
+                if (filter == FilterType::Noupdate && event == EventType::Update)
+                    || (filter == FilterType::Nodelete && event == EventType::Delete)
+                {
+                    continue;
+                }
+                let events = event_maps.get_mut(&watcher_id);
+                match events {
+                    Some(events) => {
+                        events.push(Event {
+                            key: kv.key.clone(),
+                            event: kv.event.into(),
+                            current: current.clone(),
+                            prev: prev.clone(),
+                        });
+                    }
+                    None => {
+                        let events = vec![Event {
+                            key: kv.key.clone(),
+                            event: kv.event.into(),
+                            current: current.clone(),
+                            prev: prev.clone(),
+                        }];
+                        event_maps.insert(watcher_id, events);
+                    }
+                }
+            }
+        }
+
+        // then send events to watchers
+        for (watcher_id, events) in event_maps.iter() {
+            if let Some(stream) = self.watcher_streams.get(watcher_id) {
+                let resp = WatchResponse {
+                    watch_id: *watcher_id,
+                    created: false,
+                    canceled: false,
+                    events: events.to_vec(),
+                };
+
+                stream.send(resp).await;
+            }
+        }
+
+        self.kv_vec = Vec::new();
+    }
+
+    fn recv_kv(&mut self, kv: StateMachineKvData) {
+        self.kv_vec.push(kv);
+    }
+
+    #[tracing::instrument(level = "debug", skip(self))]
+    pub async fn create_watcher_stream(
+        &mut self,
+        mut stream: Streaming<WatchRequest>,
+        tx: WatcherStreamSender,
+    ) {
+        let req = stream.message().await.unwrap().unwrap();
+
+        if let Some(CreateRequest(create)) = req.request_union {
+            tracing::info!("create_watcher_stream: {:?}", create);
+
+            let watcher_stream = WatcherStream::spawn(
+                self.current_stream_id,
+                stream,
+                tx,
+                self.watch_tx.clone(),
+                self.close_stream_tx.clone(),
+            );
+
+            self.watcher_streams
+                .insert(self.current_stream_id, watcher_stream);
+
+            self.current_stream_id += 1;
+
+            self.create_watcher(self.current_stream_id, create);
+        }
+    }
+
+    async fn watch_request(&mut self, stream_id: WatcherStreamId, req: WatchRequest) {
```

Comment:
We should report `create` and `cancel` metrics here.

Reply:
maybe metrics can be added next pr?